### DISCUSSION TITLE
How do you customize bond truncation dimension?

### DISCUSSION THREAD

**Discussion by Sinestro38** [NVIDIA/cuQuantum#105] [Q&A]
I'm trying to get the `Z^N` expectation (1 operator) of a 38 qubit GHZ state to benchmark cutensornet and I'm curious how one can customize the bond truncation dimension. By default I'm finding that the simulator is producing perfect results under 33 qubits (32 qubits is the size of the max statevec I can fit on my 40GB A100) and I suspect it's because it's not making a TN approximation.

I understand that there's an optional `SlicerOptions` object that I can pass into `OptimizerOptions` which can be passed into the `contract` function, but I couldn't find the documentation that specifies its fields. The source code also seemed to generate the dataclass from another function so thought I'd ask here.

The source code I was looking at: https://github.com/NVIDIA/cuQuantum/blob/main/python/cuquantum/cutensornet/configuration.py#L101

**Comment by DmitryLyakh** [comment]
Hi @Sinestro38 , I am not sure what you mean by bond truncation dimension in this context. Per my understanding, you are constructing your own tensor network which is supposed to compute the expectation value of the ZZ...ZZ operator over a GHZ circuit. In this case, there is no truncation and your tensor network should be evaluated with no issues. Alternatively, with the new cuTensorNet-23.10 high-level API, you can compute expectation values directly via the library API without doing any manual tensor network construction. An example can be found here: https://github.com/NVIDIA/cuQuantum/blob/main/python/samples/cutensornet/high_level/expectation_example.py

**Comment by DmitryLyakh** [comment]
If you could share your code snippet, that would help as well.

**Comment by yangcal** [comment]
Hello,

For `SlicerOption` documentation, see [here](https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.SlicerOptions.html?highlight=sliceroption).

For the old issue you were seeing (maximum dimension exceeding 32), can you make sure the quantity you're computing (expectation) is a scalar not a high dimensional tensor? `cupy` and `numpy` can't store tensors with rank exceeding 32. This is independent of the memory size.

**Reply by DmitryLyakh** [reply]
Right, computing the full state vector in order to evaluate an expectation value of some operator is not the preferred approach when using tensor network formalism for larger qubit counts.
